File: backend/src/routers/contracts.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, or_

from src.dependencies import get_db, get_current_user
from src.models import Contract, User
from src.schemas import ContractResponse

from src.services.contract_service import ContractService
from src.schemas.contract import ContractCreate
from src.schemas.review import ReviewCreate

router = APIRouter()


@router.post("/hire")
async def hire_freelancer(
    request: ContractCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role.value != "Client":
        raise HTTPException(status_code=403, detail="Only clients can hire")

    service = ContractService(db)
    contract = await service.hire_freelancer(
        job_id=request.job_id,
        bid_id=request.bid_id,
        client_id=current_user.id,
    )
    return {
        "id": contract.id,
        "job_id": contract.job_id,
        "client_id": contract.client_id,
        "freelancer_id": contract.freelancer_id,
        "bid_id": contract.bid_id,
        "amount": contract.amount,
        "status": contract.status.value,
        "start_date": contract.start_date,
    }


@router.get("")
async def get_contracts(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[ContractResponse]:
    query = select(Contract).where(
        or_(
            Contract.client_id == current_user.id,
            Contract.freelancer_id == current_user.id,
        )
    )
    result = await db.execute(query)
    contracts = result.scalars().all()

    return [
        ContractResponse(
            id=c.id,
            job_id=c.job_id,
            client_id=c.client_id,
            freelancer_id=c.freelancer_id,
            bid_id=c.bid_id,
            amount=c.amount,
            status=c.status.value,
            start_date=c.start_date,
            end_date=c.end_date,
            created_at=c.created_at,
        )
        for c in contracts
    ]


@router.get("/{contract_id}")
async def get_contract(
    contract_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.get(Contract, contract_id)
    if not result:
        raise HTTPException(status_code=404, detail="Contract not found")
    if result.client_id != current_user.id and result.freelancer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your contract")
    return result


@router.put("/{contract_id}/complete")
async def complete_contract(
    contract_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Mark a contract as complete. Only the client can call this."""
    service = ContractService(db)
    contract = await service.complete_contract(
        contract_id=contract_id,
        client_id=current_user.id,
    )
    return ContractResponse(
        id=contract.id,
        job_id=contract.job_id,
        client_id=contract.client_id,
        freelancer_id=contract.freelancer_id,
        bid_id=contract.bid_id,
        amount=contract.amount,
        status=contract.status.value,
        start_date=contract.start_date,
        end_date=contract.end_date,
        created_at=contract.created_at,
    )


@router.post("/{contract_id}/review")
async def leave_review(
    contract_id: UUID,
    request: ReviewSubmit,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Submit a review for a completed contract."""
    contract = await db.get(Contract, contract_id)
    if not contract:
        raise HTTPException(status_code=404, detail="Contract not found")
    if contract.client_id != current_user.id and contract.freelancer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your contract")

    # Determine review type automatically from caller's role
    from src.models.review import ReviewType
    role = current_user.role.value
    review_type = ReviewType.CLIENT_REVIEW if role == "Client" else ReviewType.FREELANCER_REVIEW
    reviewee_id = (
        contract.freelancer_id if role == "Client" else contract.client_id
    )

    service = ContractService(db)
    await service.leave_review(
        contract_id=contract_id,
        rating=request.rating,
        comment=request.comment,
        review_type=review_type.value,
        reviewer_id=current_user.id,
        reviewee_id=reviewee_id,
    )
    return {"detail": "Review submitted successfully"}


@router.patch("/{contract_id}")
async def update_contract(
    contract_id: UUID,
    request: dict,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.get(Contract, contract_id)
    if not result:
        raise HTTPException(status_code=404, detail="Contract not found")
    if result.client_id != current_user.id and result.freelancer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your contract")

    for key, value in request.items():
        if key == "status" and value == "Completed":
            from datetime import datetime
            result.end_date = datetime.utcnow()
        setattr(result, key, value)

    await db.flush()
    await db.refresh(result)
    return result

File: frontend/src/services/api/contracts.ts
import { axiosInstance } from './axiosInstance';
import { Contract } from '@/types';
import { MOCK_CONTRACTS } from './mockData';

const IS_MOCK = import.meta.env.VITE_MOCK_MODE === 'true';

export const contractsService = {
  getContracts: async (): Promise<Contract[]> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 800));
      return MOCK_CONTRACTS;
    }
    const { data } = await axiosInstance.get('/contracts');
    return data;
  },

  getContractById: async (id: string): Promise<Contract> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 500));
      const contract = MOCK_CONTRACTS.find(c => c.id === id);
      if (!contract) throw new Error('Contract not found');
      return contract;
    }
    const { data } = await axiosInstance.get(`/contracts/${id}`);
    return data;
  },

  hireFreelancer: async (jobId: string, bidId: string): Promise<Contract> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 1000));
      return {
        id: Math.random().toString(36).substr(2, 9),
        jobId,
        clientId: 'client-1',
        freelancerId: 'freelancer-1',
        jobTitle: 'Test Job',
        amount: 500,
        status: 'Active',
        startDate: new Date().toISOString(),
      } as Contract;
    }
    const { data } = await axiosInstance.post(`/jobs/${jobId}/hire/${bidId}`);
    return data;
  },

  completeContract: async (id: string): Promise<Contract> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 500));
      return { id, status: 'Completed' } as Contract;
    }
    const { data } = await axiosInstance.put(`/contracts/${id}/complete`);
    return data;
  },

  leaveReview: async (contractId: string, reviewData: {
    rating: number;
    comment: string;
    communication_rating?: number;
    quality_rating?: number;
    timeline_rating?: number;
  }): Promise<any> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 800));
      return { id: Math.random().toString(36).substr(2, 9), ...reviewData };
    }
    const { data } = await axiosInstance.post(`/contracts/${contractId}/review`, reviewData);
    return data;
  },

  updateContractStatus: async (id: string, status: Contract['status']): Promise<Contract> => {
    if (IS_MOCK) {
      await new Promise(resolve => setTimeout(resolve, 500));
      const contract = MOCK_CONTRACTS.find(c => c.id === id);
      if (!contract) throw new Error('Contract not found');
      return { ...contract, status };
    }
    const { data } = await axiosInstance.patch(`/contracts/${id}`, { status });
    return data;
  }
};
